### DISCUSSION TITLE
About github api rate limit

### DISCUSSION THREAD

**Discussion by Rickyxrc** [badges/shields#8339] [Q&A] [answered]
Your introduction says you serve over 870 million images a month,but github api have rate limit:5000 requests per hour for any user.Is this a problem that bothers you in your development?

**Comment by chris48s** [comment, marked as answer]
The Short answer is GitHub allows 5,000 requests per hour for any token. We increase our rate limit by using multiple tokens.

In slightly more detail: Any shields user can 'donate' us some of their rate limit by authorizing this OAuth application: https://img.shields.io/github-auth Authorizing that application gives us a token which is only authorized to access public repo content and we add it to our pool. We rarely promote this now as over the years thousands of users have authorized the app so we have enough tokens to cover our usage many times over.

**Comment by calebcartwright** [comment]
I will also add a few things:

* We're probably doing north of that 870 million number, it's something we update fairly infrequently
* We serve a lot of GitHub-affiliated badges, but not all of those 870 million+ badges are GitHub-affiliated, and thus all do not involve hits on the GitHub API
* The solution we employ as described by Chris was something that was developed years ago as part of a collaboration with GitHub folks (I'm not sure of all the specifics as that predates my time with the project)
* The authenticated rate limit has no impact/relevant on _development_, in the sense of the inner dev loop for maintainers/contributors/etc.. It only becomes for the hosted SaaS Shields.io service since that's a single badge server deployment handling a massive scale. Devs running Shields locally, or even in self-hosted Shields servers will rarely, if ever, run into the fairly gracious limits GitHub provide for a single token
